acme/challenge.go

```diff
@@ -344,6 +358,148 @@ func (hc *http01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo valida
 	return upd, nil
 }
 
+type tlsALPN01Challenge struct {
+	*baseChallenge
+}
+
+// newTLSALPN01Challenge returns a new acme tls-alpn-01 challenge.
+func newTLSALPN01Challenge(db nosql.DB, ops ChallengeOptions) (challenge, error) {
+	bc, err := newBaseChallenge(ops.AccountID, ops.AuthzID)
+	if err != nil {
+		return nil, err
+	}
+	bc.Type = "tls-alpn-01"
+	bc.Value = ops.Identifier.Value
+
+	hc := &tlsALPN01Challenge{bc}
+	if err := hc.save(db, nil); err != nil {
+		return nil, err
+	}
+	return hc, nil
+}
+
+func (tc *tlsALPN01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo validateOptions) (challenge, error) {
+	// If already valid or invalid then return without performing validation.
+	if tc.getStatus() == StatusValid || tc.getStatus() == StatusInvalid {
+		return tc, nil
+	}
+
+	config := &tls.Config{
+		NextProtos:         []string{"acme-tls/1"},
+		ServerName:         tc.Value,
+		InsecureSkipVerify: true, // we expect a self-signed challenge certificate
+	}
+
+	hostPort := net.JoinHostPort(tc.Value, "443")
+
+	conn, err := vo.tlsDial("tcp", hostPort, config)
+	if err != nil {
+		if err = tc.storeError(db,
+			ConnectionErr(errors.Wrapf(err, "error doing TLS dial for %s", hostPort))); err != nil {
+			return nil, err
+		}
+		return tc, nil
+	}
+	defer conn.Close()
+
+	cs := conn.ConnectionState()
+	certs := cs.PeerCertificates
+
+	if len(certs) == 0 {
```

Comment:
Ok this is mildly frustrating - I don't want to remove this check but I don't see a way to trigger it in tests using the HTTPS server from the standard library... Any ideas welcome

Reply:
@ibrt I don't think there's a way to trigger this in tests using the standard server, you might need to go very low level, and the tls client might crash before getting here.
Right now you can configure `tlsDial`, it's very possible that returning just a  `&tls.Conn{}` or perhaps a closed connection might do it. In any case, it's better to be safe always, so I wouldn't remove this check.

Reply:
The latter was a good hint - I got it to work thank you.